File: webapp/src/services/passcode/passcodeService.ts
/**
 * PasscodeService — App-lock passcode, COMPLETELY INDEPENDENT from encryption PIN.
 *
 * - Passcode only controls access to the UI (app lock).
 * - Passcode does NOT derive encryption keys.
 * - Passcode does NOT affect Google Drive sync or finance.json.
 * - Passcode is stored as a PBKDF2 hash (never plaintext).
 *
 * Storage keys (localStorage):
 *   pdp_passcode_enabled  — '1' when passcode is active
 *   pdp_passcode_verify   — JSON { salt: b64, iterations: number, hash: b64 }
 */

const ENABLED_KEY = 'pdp_passcode_enabled';
const VERIFY_KEY = 'pdp_passcode_verify';
const ITERATIONS = 100000; // lighter than encryption PIN (no key derivation needed)

interface VerifyToken { salt: string; iterations: number; hash: string; }

function bufToB64(buf: ArrayBuffer | Uint8Array): string {
  const bytes = buf instanceof Uint8Array ? buf : new Uint8Array(buf);
  let bin = '';
  for (let i = 0; i < bytes.length; i++) bin += String.fromCharCode(bytes[i]);
  return btoa(bin);
}
function b64ToBytes(b64: string): Uint8Array {
  const bin = atob(b64);
  const bytes = new Uint8Array(bin.length);
  for (let i = 0; i < bin.length; i++) bytes[i] = bin.charCodeAt(i);
  return bytes;
}

async function hashPasscode(passcode: string, salt: Uint8Array): Promise<string> {
  const enc = new TextEncoder();
  const baseKey = await crypto.subtle.importKey('raw', enc.encode(passcode), 'PBKDF2', false, ['deriveBits']);
  const bits = await crypto.subtle.deriveBits(
    { name: 'PBKDF2', salt, iterations: ITERATIONS, hash: 'SHA-256' },
    baseKey, 256
  );
  return bufToB64(bits);
}

class PasscodeService {
  private unlocked = false;

  isEnabled(): boolean {
    try { return localStorage.getItem(ENABLED_KEY) === '1'; } catch { return false; }
  }

  isUnlocked(): boolean {
    return !this.isEnabled() || this.unlocked;
  }

  isLocked(): boolean {
    return this.isEnabled() && !this.unlocked;
  }

  unlock(): void { this.unlocked = true; }

  lock(): void { this.unlocked = false; }

  async setup(passcode: string): Promise<void> {
    const salt = crypto.getRandomValues(new Uint8Array(16));
    const hash = await hashPasscode(passcode, salt);
    const token: VerifyToken = { salt: bufToB64(salt), iterations: ITERATIONS, hash };
    localStorage.setItem(VERIFY_KEY, JSON.stringify(token));
    localStorage.setItem(ENABLED_KEY, '1');
    this.unlocked = true;
  }

  async verify(passcode: string): Promise<boolean> {
    const raw = localStorage.getItem(VERIFY_KEY);
    if (!raw) return false;
    try {
      const token = JSON.parse(raw) as VerifyToken;
      const hash = await hashPasscode(passcode, b64ToBytes(token.salt));
      if (hash === token.hash) { this.unlocked = true; return true; }
      return false;
    } catch { return false; }
  }

  async change(oldPasscode: string, newPasscode: string): Promise<boolean> {
    if (!(await this.verify(oldPasscode))) return false;
    await this.setup(newPasscode);
    return true;
  }

  async disable(passcode: string): Promise<boolean> {
    if (!(await this.verify(passcode))) return false;
    localStorage.removeItem(ENABLED_KEY);
    localStorage.removeItem(VERIFY_KEY);
    this.unlocked = true;
    return true;
  }
}

export const passcodeService = new PasscodeService();

File: lib/services/passcode_service.dart
import 'dart:convert';
import 'dart:math';
import 'dart:typed_data';
import 'package:cryptography/cryptography.dart';
import 'package:shared_preferences/shared_preferences.dart';

/// PasscodeService — App-lock passcode, COMPLETELY INDEPENDENT from encryption PIN.
/// Passcode only controls access to the UI.
/// Passcode does NOT derive encryption keys or affect Google Drive sync.
/// Stored as PBKDF2 hash (never plaintext).
class PasscodeService {
  PasscodeService._();
  static final PasscodeService instance = PasscodeService._();

  static const String _enabledKey = 'pdp_passcode_enabled';
  static const String _verifyKey = 'pdp_passcode_verify';
  static const int _iterations = 100000;

  final Pbkdf2 _pbkdf2 = Pbkdf2(macAlgorithm: Hmac.sha256(), iterations: _iterations, bits: 256);

  Future<bool> isEnabled() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getBool(_enabledKey) ?? false;
  }

  Future<bool> verify(String passcode) async {
    final prefs = await SharedPreferences.getInstance();
    final raw = prefs.getString(_verifyKey);
    if (raw == null) return false;
    try {
      final token = jsonDecode(raw) as Map<String, dynamic>;
      final salt = Uint8List.fromList(base64.decode(token['salt'] as String));
      final hash = await _hash(passcode, salt);
      return hash == token['hash'];
    } catch (_) { return false; }
  }

  Future<void> setup(String passcode) async {
    final salt = _randomBytes(16);
    final hash = await _hash(passcode, salt);
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(_verifyKey, jsonEncode({ 'salt': base64.encode(salt), 'iterations': _iterations, 'hash': hash }));
    await prefs.setBool(_enabledKey, true);
  }

  Future<bool> change(String oldPasscode, String newPasscode) async {
    if (!await verify(oldPasscode)) return false;
    await setup(newPasscode);
    return true;
  }

  Future<bool> disable(String passcode) async {
    if (!await verify(passcode)) return false;
    final prefs = await SharedPreferences.getInstance();
    await prefs.remove(_enabledKey);
    await prefs.remove(_verifyKey);
    return true;
  }

  Future<String> _hash(String passcode, Uint8List salt) async {
    final key = await _pbkdf2.deriveKey(secretKey: SecretKey(utf8.encode(passcode)), nonce: salt);
    final bytes = await key.extractBytes();
    return base64.encode(bytes);
  }

  Uint8List _randomBytes(int n) {
    final rnd = Random.secure();
    return Uint8List.fromList(List.generate(n, (_) => rnd.nextInt(256)));
  }
}
